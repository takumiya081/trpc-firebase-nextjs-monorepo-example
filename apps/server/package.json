{
  "name": "server",
  "version": "0.0.0",
  "private": true,
  "main": "./src/index.ts",
  "types": "./src/index.ts",
  "scripts": {
    "test": "vitest"
  },
  "devDependencies": {
    "@packages/db": "workspace:*",
    "@types/node": "^17.0.12",
    "@types/react-dom": "^18.0.7",
    "eslint": "7.32.0",
    "eslint-config-custom": "workspace:*",
    "tsconfig": "workspace:*",
    "typescript": "^4.5.3"
  },
  "dependencies": {
    "@packages/db": "workspace:*",
    "@trpc/server": "^10.9.1",
    "bson": "^5.0.0",
    "dotenv": "^16.0.3",
    "dotenv-expand": "^10.0.0",
    "firebase-admin": "11.5.0",
    "mongodb": "^5.0.0"
  }
}
