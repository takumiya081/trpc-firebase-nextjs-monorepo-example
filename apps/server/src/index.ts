export { appRouter, type AppRouter } from "./root";
export { createTRPCContext } from "./trpc";
import superjson from 'superjson';

export const transformer = superjson;