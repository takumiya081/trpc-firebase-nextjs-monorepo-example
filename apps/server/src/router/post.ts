import { z } from "zod";

import { createTRPCRouter, publicProcedure, protectedProcedure } from "../trpc";
import type {} from '@packages/db';
import { ObjectId } from "bson";

export const postRouter = createTRPCRouter({
  all: publicProcedure.query(({ ctx }) => {
    return ctx.prisma.post.findMany({ orderBy: { id: "desc" } });
  }),
  byId: publicProcedure.input(z.object({id: z.string()})).query(({ ctx, input }) => {
    return ctx.prisma.post.findFirst({ where: { id: input.id } });
  }),
  delete: publicProcedure.input(z.string()).mutation(({ ctx, input }) => {
    return ctx.prisma.post.delete({ where: { id: input } });
  }),
  update: protectedProcedure.input(z.object({id: z.string(), content: z.string().optional(), title: z.string().optional()})).mutation(({ctx, input}) => {
    return ctx.prisma.post.update({
      where: {
        id: input.id,
      },
      data: {
        content: input.content,
        title: input.title,
      }
    })
  }),
  create: protectedProcedure.input(z.object({content: z.string(), title: z.string()})).mutation(({ctx, input}) => {
    return ctx.prisma.post.create({
      data: {
        id:new ObjectId().toHexString(),
        content: input.content,
        title: input.title,
      }
    })
  })
});