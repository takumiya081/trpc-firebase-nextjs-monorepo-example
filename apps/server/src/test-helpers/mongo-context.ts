import {MongoClient} from 'mongodb';

import type {PrismaClient} from '@packages/db';
import {testSetUp} from '@packages/db';

export class MongoContext {
  private databaseName?: string;

  private databaseUrl?: string;

  private prismaClientInstance: undefined | PrismaClient = undefined;

  private clear?: () => Promise<void>;

  public get prismaClient(): PrismaClient {
    if (this.prismaClientInstance === undefined) {
      throw new Error('prismaClient is undefined');
    }
    return this.prismaClientInstance;
  }

  public async before() {
    try {
      const testHelpers = await testSetUp();
      this.databaseName = testHelpers.databaseName;
      this.databaseUrl = testHelpers.databaseUrl;
      this.prismaClientInstance = testHelpers.prisma;
      this.clear = testHelpers.clear;
    } catch (e) {
      // eslint-disable-next-line no-console
      console.log(`error on push`, e);
    }
  }

  public async after() {
    try {
      if (!this.databaseUrl) {
        throw new Error(`this.databaseUrl is undefined`);
      }
      const client = await new MongoClient(this.databaseUrl).connect();
      await client.db(this.databaseName).dropDatabase();
      await client.close();
      await this.clear?.();
    } catch (e) {
      // eslint-disable-next-line no-console
      console.log('delete db error', e);
    }
  }
}
