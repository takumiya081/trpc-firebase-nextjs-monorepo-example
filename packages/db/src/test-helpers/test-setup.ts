import {PrismaClient} from '@prisma/client';
import {execSync} from 'child_process';
import fs from 'fs';
import path from 'path';

export function createTestPrismaClient(databaseUrl: string) {
  return new PrismaClient({datasources: {db: {url: databaseUrl}}});
}

function generateDatabaseName() {
  const r = Math.random().toString(36).slice(2, 7);
  return r;
}

function getPrismaBinary() {
  let currDir = __dirname;
  // eslint-disable-next-line no-constant-condition
  while (true) {
    const binaryPath = path.join(currDir, 'node_modules/.bin/prisma');
    if (fs.existsSync(binaryPath)) {
      return binaryPath;
    }
    const {dir, root} = path.parse(currDir);
    if (dir === root) {
      throw new Error(
        `Could not find prisma binary in the parent directories starting from ${__dirname}.`,
      );
    }
    currDir = dir;
  }
}

function getPrismaSchemaPath() {
  let currDir = __dirname;
  // eslint-disable-next-line no-constant-condition
  while (true) {
    const binaryPath = path.join(currDir, 'prisma/schema.prisma');
    if (fs.existsSync(binaryPath)) {
      return binaryPath;
    }
    const {dir, root} = path.parse(currDir);
    if (dir === root) {
      throw new Error(
        `Could not find prisma/schema.prisma in the parent directories starting from ${__dirname}.`,
      );
    }
    currDir = dir;
  }
}

function createTestDatabaseUrl(dbName: string) {
  // url format like `mongodb://root:password@localhost:27018/tests?authSource=admin`
  const query = process.env.DATABASE_URL.match(/\?.+/gi);
  return process.env.DATABASE_URL.replace(
    /^(mongodb:\/\/.+\/).+(\?.*)?$/gi,
    `$1${dbName}${query ? query[0] : ''}`,
  );
}

export async function testSetUp() {
  const databaseName = generateDatabaseName();
  const databaseUrl = createTestDatabaseUrl(databaseName);

  const prismaBinary = getPrismaBinary();
  const prismaSchemaPath = getPrismaSchemaPath();
  execSync(`${prismaBinary} db push --skip-generate --schema ${prismaSchemaPath}`, {
    env: {...process.env, DATABASE_URL: databaseUrl},
  });

  const prismaClientInstance = new PrismaClient({datasources: {db: {url: databaseUrl}}});

  async function clear() {
    await prismaClientInstance.$disconnect();
  }

  return {
    prisma: prismaClientInstance,
    databaseUrl,
    databaseName,
    clear,
  };
}
