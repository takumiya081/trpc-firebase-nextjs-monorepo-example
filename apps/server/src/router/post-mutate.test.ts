import {MongoContext} from '~/test-helpers';
import {expect, beforeAll, describe, afterAll, test} from 'vitest'

import {postRouter} from './post'
import { inferProcedureInput } from '@trpc/server';
import { AppRouter } from '..';
import {ObjectId} from 'bson'

describe('public module query', () => {
  const mongoContext = new MongoContext();
  const savedPost = {
    id: new ObjectId().toHexString(),
    title: 'test',
    content: 'test'
  }

  beforeAll(async () => {
    await mongoContext.before();

    await mongoContext.prismaClient.post.createMany({data: [
      savedPost
    ]})
  });

  afterAll(async () => {
    await mongoContext.after();
  });

  test('should success on create when authed', async () => {
    expect.hasAssertions();
    const caller = postRouter.createCaller({prisma: mongoContext.prismaClient, user: {id: 'foo'} as any})


    const input: inferProcedureInput<AppRouter['post']['create']> = {
      title: 'test title',
      content: 'test content',
    };

    const res = await caller.create(input);
    expect(res).toMatchSnapshot({id: expect.any(String)});
  })

  test('should success on update when authed', async () => {
    expect.hasAssertions();
    const caller = postRouter.createCaller({prisma: mongoContext.prismaClient, user: {id: 'foo'} as any})


    const input: inferProcedureInput<AppRouter['post']['update']> = {
      id: savedPost.id,
      title: 'updated title',
      content: 'updated content',
    };

    const res = await caller.update(input);
    expect(res).toMatchSnapshot({id: expect.any(String)}, 'res');
  })
});