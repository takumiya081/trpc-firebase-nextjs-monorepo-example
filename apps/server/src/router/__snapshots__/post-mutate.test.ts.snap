// Vitest Snapshot v1

exports[`public module query > should success on create when authed 1`] = `
{
  "content": "test content",
  "id": Any<String>,
  "title": "test title",
}
`;

exports[`public module query > should success on update when authed > res 1`] = `
{
  "content": "updated content",
  "id": Any<String>,
  "title": "updated title",
}
`;
